### ISSUE TITLE
JsonSerializer.Serialize call doesn't return and continues to run for large lists.

### ISSUE THREAD

**Issue by msucgun** [dotnet/runtime#73992]
### Description

Hi,

I have a list that includes 2.000.000 items. While calling JsonSerializer.Serialize method for it, the method call keeps running recursively and doesn't return any result. The memory usage pattern during the method call is as follows.

![vs_dbg_cpu_ram](https://user-images.githubusercontent.com/929462/184800626-979fc021-2a5b-4b3f-a8af-8104018ce3e4.png)

### Reproduction Steps

I've created a project to reproduce the same behavior. You can get it from [https://github.com/msucgun/poc/tree/main/JsonSerializeMethodNotReturn](https://github.com/msucgun/poc/tree/main/JsonSerializeMethodNotReturn).

### Expected behavior

The method call should return a result or throw an exception if there is any limitation for the serialization operation and that scenario exceeds the limits.

### Actual behavior

The method call keeps running recursively and doesn't return any result.

### Regression?

In 3.1 runtime, the method call thrown a System.OverflowException.

### Known Workarounds

Splitting the list into chunks and serializing them separately works, then that's not a choice for us.

### Configuration

Runtime: 6.0.7
SDK: 6.0.302
OS: Windows 10 (Version 20H2 Build 19042.1826)
Architecture: Any CPU (x86/x64)
It's not specific to that configuration. The same issue occurs on Windows Server 2016 (Version 1607 Build 14393.5246) with runtime 6.0.5.

**Comment by steveharter** [comment]
The repro is missing the "Faker" class.

Working around that, I can serialize 2,000,000 items without error. However, soon after that it looks like `PooledByteBufferWriter` is using `int` so it will overflow and does not work correctly. Either there should be some validation to prevent that (I thought there was at some point) or it should support it by using `long\ulong`.

**Comment by msucgun** [comment]
> The repro is missing the "Faker" class.
> 
> Working around that, I can serialize 2,000,000 items without error. However, soon after that it looks like `PooledByteBufferWriter` is using `int` so it will overflow and does not work correctly. Either there should be some validation to prevent that (I thought there was at some point) or it should support it by using `long\ulong`.

@steveharter thanks for the reply. Faker class comes from the package Bogus. Have you cloned the whole repo from [https://github.com/msucgun/poc](https://github.com/msucgun/poc) and restored the nuget packages? I've created the solution using VS2022 and didn't have any issue to build and run. If you run the sample in the solution (with data prepared with RandomListGenerator.GenerateDtoList method) you can see the issue.

**Comment by huoyaoyuan** [comment]
Another hint is to use `DeserializeAsyncEnumerable` to access the values before deserializing the full list.

**Comment by steveharter** [comment]
This does appear to be an issue with `PooledByteBufferWriter` hitting the 2GB limit for `int32` for the [`index` field](https://github.com/dotnet/runtime/blob/c785a0b5ff94d879c55a9a67a18da0edf1070f32/src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs#L21). The field can't simply be changed to a `long` since it uses other APIs that are based on `int32`, such as `MemoryExtensions.AsMemory(int start, int length)`.

Either this should be fixed or a proper exception thrown regarding max length exceeded.

**Comment by steveharter** [comment]
> Faker class comes from the package Bogus. Have you cloned the whole repo from https://github.com/msucgun/poc and restored the NuGet packages?

Thanks it is working now for me now.

**Comment by krwq** [comment]
I think we should at minimum throw an exception for 7.0 if fix is more involved

**Comment by eiriktsarpalis** [comment]
Does it meet the bar for a .NET 7 fix? It's not a recent regression AFAICT.

**Comment by msucgun** [comment]
> I think we should at minimum throw an exception for 7.0 if fix is more involved

In fact, I expect that .NET 7 comes with large-sized object support. Throwing an exception would be a quick-win solution and better to be scheduled for 6.0.

**Comment by steveharter** [comment]
> I think we should at minimum throw an exception for 7.0 if fix is more involved

I agree.

> Does it meet the bar for a .NET 7 fix? It's not a recent regression AFAICT.

I think so. This has not been reported before AFAIK but if it was it would have been fixed. It falls into the "infinite loop" or "data damaging" category which is not good.

**Comment by steveharter** [comment]
The exception message may want to say use the Stream\async APIs instead.

**Comment by eiriktsarpalis** [comment]
>> Does it meet the bar for a .NET 7 fix? It's not a recent regression AFAICT.
>
>I think so. This has not been reported before AFAIK but if it was it would have been fixed. It falls into the "infinite loop" or "data damaging" category which is not good.

Should we service 6.0 as well?

**Comment by krwq** [comment]
we should definitely bar check 6.0

**Comment by layomia** [comment]
In .NET Core 3.1 the serializer failed early with OverflowException due to faulty buffer growth logic. This bug has since been fixed (https://github.com/dotnet/runtime/pull/1308) and an explicit OOM is thrown instead, hence the perceived behavior change.

A “hang” here is the serializer being slow to serialize the data, but in all cases it eventually completes with OOM, not a stack or numeric overflow. I’m not able to consistently reproduce a hang, meaning that the serializer sometimes completes the operation in bounded time.

Using the default buffer size (16K, configurable with JsonSerializerOptions.DefaultBufferSize) used to create an initial buffer for writing (and upon which size doubling is based), the serializer always hangs on my machine.

Using a large enough buffer size (e.g. Array.MaxLength / 2), though perhaps impractical in real-world scenarios, the serializer sometimes completes serialization, but then throws OOM when transcoding to string using the lower UTF-8 encoding APIs (due to the contiguous input being too large).

In conclusion, my finding is that the serializer will either complete the operation or eventually throw OOM. I believe that the logic in the PoolBufferWriter is sound. We might also consider async serialization methods based on `IAsyncEnumerable` if processing of large lists is commonly desired, similar to the existing deserialization APIs.

**Comment by layomia** [comment]
We might want to also consider a quality of life improvement to throw an exception sooner.

**Comment by steveharter** [comment]
When I tested locally, performance was normal or as-expected until the 2GB was approached, then performance tanked and appeared to hang (I didn't wait too long). @layomia did you encounter this as well?

**Comment by steveharter** [comment]
I suspect the algorithm, once past 1GB, is causing the buffer to only grow a small amount (the size needed or the min of 256 bytes) per resize until it hits 2GB. See https://github.com/dotnet/runtime/blob/c785a0b5ff94d879c55a9a67a18da0edf1070f32/src/libraries/Common/src/System/Text/Json/PooledByteBufferWriter.cs#L173-L183.

If instead it detects size is already at 1GB or more, it could change the "MinimumBufferSize" to 100K or even allocate the full 2GB to avoid the perf issue.

**Comment by layomia** [comment]
@steveharter yes I did observe a sharp perf tank as serialization went on. Your suggestion to allocate the full buffer after reaching halfway fixed the slow perf. In the reported issue, we reach the OOM very quickly now with the PR fix - https://github.com/dotnet/runtime/pull/75147.
